tests/integration/test_coredns.py

```diff
@@ -29,41 +27,55 @@ async def test_core_dns(e: OpsTest):
         )
         await e.model.wait_for_idle(["microk8s"])
 
-    u: Unit = e.model.applications["microk8s"].units[0]
+    yield e.model.applications["microk8s"]
+
 
+@pytest.fixture()
+async def coredns(e: OpsTest, microk8s):
     # bootstrap a juju cloud on the deployed microk8s
-    async with microk8s_kubernetes_cloud_and_model(e, "microk8s") as (k8s_model, model_name):
-        with e.model_context(k8s_model):
+    async with microk8s_kubernetes_cloud_and_model(e, "microk8s") as (k8s_alias, _):
+        with e.model_context(k8s_alias) as k8s_model:
             LOG.info("Deploy CoreDNS")
-            await e.model.deploy(
+            await k8s_model.deploy(
                 config.MK8S_CORE_DNS_CHARM,
                 application_name="coredns",
                 channel=config.MK8S_CORE_DNS_CHANNEL,
                 trust=True,
             )
-            await e.model.wait_for_idle(["coredns"])
-
             LOG.info("Create offer for dns-provider")
-            await e.model.create_offer("coredns:dns-provider", "coredns")
+            await k8s_model.create_offer("coredns:dns-provider", "coredns")
+            await k8s_model.wait_for_idle(["coredns"])
+
+        yield k8s_model.applications["coredns"]
 
-        try:
-            LOG.info("Consume offer for dns-provider")
-            await e.model.consume(f"admin/{model_name}.coredns", "coredns")
-            LOG.info("Add relation between microk8s and coredns")
-            await e.model.add_relation("microk8s", "coredns")
-            LOG.info("Wait for idle")
-            await e.model.wait_for_idle(["microk8s"])
+        k8s_model.remove_offer("coredns")
+        k8s_model.remove_application("coredns", block_until_done=True)
 
-            with e.model_context(k8s_model):
-                await e.model.wait_for_idle(["coredns"])
 
-            rc = None
-            while rc != 0:
-                rc, stdout, stderr = await run_unit(u, f"microk8s kubectl run -i --tty --rm debug --image=busybox --restart=Never -- nslookup google.com")
-                LOG.info("Verify the pod dns resolution %s", (rc, stdout, stderr))
+@pytest.mark.abort_on_fail
+async def test_core_dns(e: OpsTest, microk8s, coredns):
```

Comment:
test by consuming the offer in the machine_model, waiting for both models to be idle, and running the dns query

then tearing out the relation and saas